File: 2024/day05/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

// Parse input file and extract rules and updates
void parseInput(const string& filename,
                unordered_map<int, unordered_set<int>>& rules,
                vector<vector<int>>& updates) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Cannot open file " << filename << endl;
        exit(1);
    }

    string line;
    bool parsingRules = true;

    while (getline(file, line)) {
        if (line.empty()) {
            parsingRules = false;
            continue;
        }

        if (parsingRules) {
            // Parse rule: X|Y
            size_t pos = line.find('|');
            int before = stoi(line.substr(0, pos));
            int after = stoi(line.substr(pos + 1));
            rules[before].insert(after);
        } else {
            // Parse update: comma-separated numbers
            vector<int> update;
            stringstream ss(line);
            string num;
            while (getline(ss, num, ',')) {
                update.push_back(stoi(num));
            }
            updates.push_back(update);
        }
    }

    file.close();
}

// Check if an update is in valid order according to rules
bool isValidOrder(const vector<int>& update,
                  const unordered_map<int, unordered_set<int>>& rules) {
    // Create a map of page positions
    unordered_map<int, int> pagePositions;
    for (int i = 0; i < update.size(); i++) {
        pagePositions[update[i]] = i;
    }

    // Check each page and its constraints
    for (int i = 0; i < update.size(); i++) {
        int page = update[i];

        // Check if this page has any ordering rules
        auto it = rules.find(page);
        if (it != rules.end()) {
            // Check all pages that must come after this page
            for (int mustBeAfter : it->second) {
                // If mustBeAfter is in this update
                auto posIt = pagePositions.find(mustBeAfter);
                if (posIt != pagePositions.end()) {
                    // It must come after current page
                    if (posIt->second < i) {
                        return false;
                    }
                }
            }
        }
    }

    return true;
}

// Fix order of an update using custom comparator
vector<int> fixOrder(const vector<int>& update,
                     const unordered_map<int, unordered_set<int>>& rules) {
    vector<int> result = update;

    // Sort using custom comparator
    sort(result.begin(), result.end(), [&rules](int a, int b) {
        // If a must come before b, return true
        auto it = rules.find(a);
        if (it != rules.end() && it->second.count(b) > 0) {
            return true;
        }

        // If b must come before a, return false
        auto it2 = rules.find(b);
        if (it2 != rules.end() && it2->second.count(a) > 0) {
            return false;
        }

        return false;
    });

    return result;
}

// Part 1: Sum middle page numbers of correctly-ordered updates
int part1(const vector<vector<int>>& updates,
          const unordered_map<int, unordered_set<int>>& rules) {
    int total = 0;

    for (const auto& update : updates) {
        if (isValidOrder(update, rules)) {
            int middleIdx = update.size() / 2;
            total += update[middleIdx];
        }
    }

    return total;
}

// Part 2: Sum middle page numbers of fixed incorrectly-ordered updates
int part2(const vector<vector<int>>& updates,
          const unordered_map<int, unordered_set<int>>& rules) {
    int total = 0;

    for (const auto& update : updates) {
        if (!isValidOrder(update, rules)) {
            vector<int> fixed = fixOrder(update, rules);
            int middleIdx = fixed.size() / 2;
            total += fixed[middleIdx];
        }
    }

    return total;
}

int main() {
    unordered_map<int, unordered_set<int>> rules;
    vector<vector<int>> updates;

    parseInput("../input.txt", rules, updates);

    cout << "Part 1: " << part1(updates, rules) << endl;
    cout << "Part 2: " << part2(updates, rules) << endl;

    return 0;
}

File: 2024/day05/clojure/solution.clj
(require '[clojure.string :as str])
(require '[clojure.java.io :as io])

;; Read and parse input
(def input-text
  (str/trim (slurp "../input.txt")))

;; Split into sections
(def sections (str/split input-text #"\n\n"))
(def rules-section (str/split-lines (first sections)))
(def updates-section (str/split-lines (second sections)))

;; Parse rules: X|Y means X must come before Y
;; Store as: rules map where rules[X] = set of pages that must come AFTER X
(def rules
  (reduce
    (fn [acc rule-line]
      (let [[before after] (map #(Integer/parseInt %) (str/split rule-line #"\|"))]
        (update acc before (fnil conj #{}) after)))
    {}
    rules-section))

;; Parse updates
(def updates
  (map (fn [line]
         (map #(Integer/parseInt %) (str/split line #",")))
       updates-section))

(defn is-valid-order?
  "Check if an update is in valid order according to rules."
  [update]
  (let [page-positions (into {} (map-indexed (fn [i page] [page i]) update))]
    (every?
      (fn [[i page]]
        (let [must-be-after (get rules page #{})]
          (every?
            (fn [after-page]
              (if-let [after-pos (get page-positions after-page)]
                (>= after-pos i)
                true))
            must-be-after)))
      (map-indexed vector update))))

(defn part1
  "Sum middle pages of correctly-ordered updates."
  []
  (reduce
    (fn [total update]
      (if (is-valid-order? update)
        (let [update-vec (vec update)
              middle-idx (quot (count update-vec) 2)]
          (+ total (nth update-vec middle-idx)))
        total))
    0
    updates))

(defn fix-order
  "Reorder an update to satisfy all rules using custom comparator."
  [update]
  (sort
    (fn [a b]
      (cond
        ;; If a must come before b
        (contains? (get rules a #{}) b) -1
        ;; If b must come before a
        (contains? (get rules b #{}) a) 1
        ;; No ordering constraint
        :else 0))
    update))

(defn part2
  "Sum middle pages of incorrectly-ordered updates after fixing."
  []
  (reduce
    (fn [total update]
      (if-not (is-valid-order? update)
        (let [fixed (fix-order update)
              fixed-vec (vec fixed)
              middle-idx (quot (count fixed-vec) 2)]
          (+ total (nth fixed-vec middle-idx)))
        total))
    0
    updates))

;; Main execution
(println (str "Part 1: " (part1)))
(println (str "Part 2: " (part2)))
